### DISCUSSION TITLE
Using Raspberry Pi as scan server?

### DISCUSSION THREAD

**Discussion by olewisddev** [alexpevzner/sane-airscan#115] [Q&A] [answered]
I was wondering whether I could use this package in a Raspberry Pi (Raspbian OS) for creating a wireless scan server. I have an **HP Deskjet Ink Advantage 1515** printer/scanner and it doesn't have a built-in network support. I am now using this Raspberry Pi as a wireless print server using CUPS but I wanted to extend its feature to wireless scanning.

Additional question: How is this package different from [SimulPiscator/AirSane](https://github.com/SimulPiscator/AirSane) besides the web-interface?

**Comment by alexpevzner** [comment, marked as answer]
The `sane-airscan` is a scanner driver for SANE, not the server.

If your printer supports IPP-over-USB protocol, you may try to use `ipp-usb` (https://github.com/OpenPrinting/ipp-usb) to re-export it to the network. Regardless of its name, IPP-over-USB typically exposes printing, scanning and device's web console.

To check if IPP-over-USB is supported by device, do `lsusb -v` as root, find your printer's entry and look for the following lines:
```
      bInterfaceClass         7 Printer
      bInterfaceSubClass      1 Printer
      bInterfaceProtocol      4
```

If you see combination of Class=7 and Protocol=4, your device supports IPP-over-USB.

The `ipp-usb` daemon by default doesn't expose devices to the network, only to `localhost`. To change this behaviour, edit `/etc/ipp-usb/ipp-usb.conf`, find the following line in the `[network]` section:
```
[network]
  interface = loopback
```
And replace `loopback` with `all`.

If it works for you, your printer will appear on a network as an IPP printer and eSCL scanner, and no additional software on the Raspberry Pi side will be needed (no CUPS etc).

The `AirSane` program does an opposite to `sane-airscan` task: it converts any SANE-compatible scanner into the eSCL server. If you will use `AirSane`, you'll probably will need `sane-airscan` on a client side, because `sane-escl` driver, included into the `sane-backends` package doesn't work with `AirSane` (at least, it didn't some time ago). And you will need to run CUPS on your Raspberry Pi to export printing parf of your device.

My recommendation: use `ipp-usb`, if your printer is supported by this program, otherwise fall back to `AirSane` or even `saned` (comes with SANE)
```

**Reply by olewisddev** [reply]
So I checked with my current setup and it seems that only the `bInterfaceProtocol` is different. I had the following relevant lines after running `lsusb -v` as a superuser:
```
bInterfaceClass       7 Printer
bInterfaceSubClass    1 Printer
bInterfaceProtocol    2 Bidirectional
```
Does this mean I cannot use `ipp-usb`? Does the value of `bInterfaceProtocol` has something to do with the CUPS server I already set up and `hplip`? I used `lsusb` while the relevant CUPS packages are still installed and running, will removing those change the value of `bInterfaceProtocol`?

If I use `AirSane` instead, will Apple devices detect the scanner automatically or do I still need to install an additional app? I think Fedora, the distribution I use, already includes `sane-airscan` natively so I can technically run GNOME Document Scanner and it will automatically search then  detect the scanner too. Is my assumption correct?

Edit: Had an additional question

**Comment by alexpevzner** [comment]
There may be multiple lines per device, but if none of them contains 7/4 combination, yes, it means that you can't use `ipp-usb` with this device.

These numbers are hardwired into the device, and software installed and running on the host can't change them.

**Reply by olewisddev** [reply]
Thank you for your prompt response. I edited my previous reply with an additional question. I hope you can also address it too.

**Comment by alexpevzner** [comment]
> If I use AirSane instead, will Apple devices detect the scanner automatically or do I still need to install an additional app?

I don't know. AirSane author claims so, but I briefly tested with Mopria for Android (it also uses eSCL), and didn't succeed. It may be my mistake though, I didn't investigate thoroughly. But `ipp-usb` on a same machine works well with Mopria.

> I think Fedora, the distribution I use, already includes sane-airscan natively

Yes, starting from Fedora 32

> so I can technically run GNOME Document Scanner and it will automatically search then detect the scanner too. Is my assumption correct?

Yes

**Comment by markosjal** [comment]
I am working on a project currently to share a windows scanner by way of eSCL. Once I get through that , it is my intention to make the SANE to eSCL equivalent for Linux. 

I have tested the Airsane package with mixed results .if I felt it were up to par, I would not bother with my own. Unfortunately with Airsane I have seen client incompatibility as well as some scanners being incompatible. I do not see it improving rapidly. 

I will say after playing with eSCL for a couple years now that building an eSCL server is much more complex than the client.
